### ISSUE TITLE
Java Mouse Location Incorrect when moving a window

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#1941]
*Issue migrated from trac ticket # 1941*

**component:** server | **priority:** critical

#### 2018-08-27 11:59:16: mjharkin created the issue

___
> 
> I have an undecorated window that I want to move by dragging the panel inside. However under the scenario WindowDragNoDelay the window jumps erratically with the mouse position read incorrectly in Java (also printed to stdout).
> 
> Through debugging I've 2 scenarios (WindowDragWithDelay,WindowDragNoMove) that can return the mouse location correctly and 1 (WindowDragNoDelay) that doesn't.
> 
> As a control (CentOS without Xpra) WindowDragNoDelay will return the correct mouse location.
> 
> Reproduce by installing java and running the windowdrag.jar in attached zip file:
> ```
> java -cp windowdrag.jar WindowDragNoDelay
> java -cp windowdrag.jar WindowDragWithDelay
> java -cp windowdrag.jar WindowDragNoMove
> ```
> 
> Can attach logs if you provide the log categories you want.
> 
> Python Client on Windows 10 (also can reproduce with HTML5 client)
> Server on CentOS 7.
> Client and server both on revision [r20214](../commit/4645db5a368227fbf0a87138ee06c3e53ff10db1)

**Comment by totaam** [comment]
Correct fix in [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9), this was hard because everything is asynchronous, including X11 events, client-server messaging, etc
So now the client keeps track of the delta between the window location requested by the server and what it actually is using at any point in time. It communicates this delta value with the server through the "map" and "configure" packets.
Then the server can adjust pointer events with the correct delta, even if the window has moved since the last time the client got a move notification.

Still TODO:
* `ClientTray` class could be updated so we can eventually replace the `mapped_at` attribute with just the delta, since the former is prone to race conditions
* update the html5 client? (same reason)
* maybe move all window scaling code to the window class (rather than passing both scaled and unscaled coordinates)
* created new ticket: #1942 to make the code more manageable in the future
* re-test: shadow servers, scaling, etc

@mjharkin: your test case now works with this new code. You will need an updated client and server (2.4 beta [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9) or later, builds have been uploaded for mswindows, centos 7.4 and 7.5: [https://xpra.org/beta/]). This change is not suitable for backporting to older branches.

Note that this is not how you should be doing the window move from your Java code, as this is excruciatingly slow, even on local connections.
Instead of managing the window position yourself (which will happen server-side and cause the churn), you should be requesting that the window manager moves the window for you. This will happen client-side and will be smooth as silk. (though somewhat less smoothly on mswindows and macos than other platforms since we have to emulate the X11 window-manager-spec's `_NET_WM_MOVERESIZE`: #772)
This can be achieved on X11 with [_NET_WM_MOVERESIZE](https://standards.freedesktop.org/wm-spec/wm-spec-1.3.html#idm140130317636976), you can find an example we use for testing here: [/browser/xpra/trunk/src/tests/xpra/test_apps/test_window_initiatemoveresize.py]

**Comment by totaam** [comment]
> `disconnect: zlib packet decompression failed must be string or read-only buffer, not memoryview`

Fixed: #1926#comment:2.

> Not sure if it's caused by these changes though.

It wasn't, but thanks for reporting it.
(out of curiosity, what made you try the websocket transport? it is slower after all)

> ` install and enable lzo or lz4 support for better performance`

For lz4, see #1929#comment:1. (patching required)
For lzo, try this from a mingw shell:

```shell
python -c "import lzo"
```
Is it installed? Or did it not get packaged properly?
(try to run `Network_info.exe -v`)

> `Error: failed to load overlay icon ...`
> (..)
> `DLL load failed: The specified module could not be found.`

Hmm. Things are not going to work well without Pillow and the codecs.
Maybe your system has a dependency that I don't have and it doesn't get packaged?
Are you fully up to date? (`pacman -Syyu`)
Try updating all the python libraries too:
```shell
sh win32/UPDATE_PYTHON_LIBS.sh
```
You can try to figure out which DLL is missing using [Dependencies](https://github.com/lucasg/Dependencies) (a new "dependency walker") by pointing it to the "pyd" files that generate the DLL warnings.
If you still have problems with this, can you follow up in #1883 or even a new ticket?
(let's keep this ticket about pointer position)


----

> [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9) introduces a serious mouse bug. I've double checked this with a bisection and [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9) is the guilty revision.

[r20230](../commit/6cb2683289ab3b7826c6340aaf666dd8e04a2460) fixes this: when the client moves a window (or maps it), the server should only apply the delta value until the client catches up (if it ever does).
Which means that this statement:
> .. so we can eventually replace the mapped_at attribute with just the delta, since the former is prone to race conditions

Is wrong. We need to keep both.
>

**Comment by totaam** [comment]
Taking the ticket back: forgot `reinit_windows` (fixed in [r20243](../commit/37ad617a598ada6f3db50a83e2698cbfc67bc653)) and this solution isn't going to work well with sharing enabled - I think I have a better plan.

**Comment by totaam** [comment]
Reverted [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9) + [r20230](../commit/6cb2683289ab3b7826c6340aaf666dd8e04a2460) in [r20249](../commit/a3abcc2b2b1e514fd5d132e485f7d9fa73694983). Still working on a correct solution..

Related: [r20252](../commit/33631bbdedb3d16fca837464f43fbb9aba141ca8) + [r20253](../commit/0baf653983506867e0ab7c99db0adc17c7046ac6) exposes window-relative pointer position in pointer-position and button-action packets, this may help and could actually simplify pointer adjustments when dealing with desktop and shadow servers: with those we never care about the position of the window on the client, only the relative coordinates are useful.

**Comment by totaam** [comment]
Too late for 2.4, this sort of change can break things.

**Comment by totaam** [comment]
#2008 could also be related to this bug.

**Comment by totaam** [comment]
Re-scheduling, might be easier to make any changes after #1942.

**Comment by totaam** [comment]
The revert was incomplete and caused a bug: #2249.

**Comment by totaam** [comment]
I was thinking that we could simply calculate the current window offset from the relative pointer coordinates, then adjust from that value. Not so easy I am afraid.

Tracking back:
* [r14368](../commit/d92799331a25069f4826e67e2181ee688ffdb012) is where the `_adjust_pointer` method was generalized, recording where each client maps the window in the `mapped_at` attribute of window source object
* [r20252](../commit/33631bbdedb3d16fca837464f43fbb9aba141ca8) added relative pointer coordinates, unfortunately this also added the `_get_pointer_abs_coordinates` method which means the disable-pointer-adjustment.patch "fix" no longer works - not sure why we need that: adjustments, if any are needed, should be in one place? Maybe this is for multi-window shadow servers?

So, we do want to use `adjust_pointer` when the window is mapped at a different location (ie: because the client's OS prevented it from overflowing), but not in other cases (ie: when we've just moved it and the client hasn't caught up yet...)

**Comment by totaam** [comment]
Damn. I was thinking of using another approach and let the client do all the adjustments to pointer coordinates, returning:
* absolute position
* window relative position
* and adding: window relative to what the server told us to use

It would be easier because then there would only be limited logic on the server side, just using the server-relative coordinates when those values are provided by the client.
The problem with this approach is that the client doesn't always know where the window is actually mapped on the server side: it could have been moved since, and when it sends a "configure" or "map" packet, the server may or may not honour it: only the server knows the actual window position to do the adjustments.

So the only viable solution is going to be something like [r20225](../commit/9c9727dc6f99f60edd6458692c7dbc96286f88d9) + [r20230](../commit/6cb2683289ab3b7826c6340aaf666dd8e04a2460) + [r20243](../commit/37ad617a598ada6f3db50a83e2698cbfc67bc653).

**Comment by totaam** [comment]
See also #2600, #2516, #2618, #2675, #2455. Raising.

**Comment by totaam** [comment]
> FYI the zip file seems broken

Looks fine to me:
```
$ unzip WinDrag.zip 
Archive:  WinDrag.zip
   creating: src/
(..)
  inflating: .classpath              
```

**Comment by totaam** [comment]
See also https://github.com/Xpra-org/xpra/issues/3938#issuecomment-1939154129
